presto-main/src/main/java/com/facebook/presto/operator/GenericPageProcessor.java

```diff
@@ -36,16 +51,153 @@ public GenericPageProcessor(FilterFunction filterFunction, Iterable<? extends Pr
+
+        for (int projectionIndex = 0; projectionIndex < projections.size(); projectionIndex++) {
+            ProjectionFunction projection = projections.get(projectionIndex);
+            BlockBuilder blockBuilder = pageBuilder.getBlockBuilder(projectionIndex);
+            for (int position : selectedPositions) {
+                projection.project(position, inputBlocks, blockBuilder);
+            }
+        }
+        pageBuilder.declarePositions(selectedPositions.length);
+        return pageBuilder.build();
+    }
+
+    @Override
+    public Page processColumnarDictionary(ConnectorSession session, Page page, List<? extends Type> types)
+    {
+        Page inputPage = getNonLazyPage(page);
+        int[] selectedPositions = filterPage(inputPage);
+
+        if (selectedPositions.length == 0) {
+            return null;
+        }
+
+        if (projections.isEmpty()) {
+            return new Page(selectedPositions.length);
+        }
+
+        PageBuilder pageBuilder = new PageBuilder(types);
+        Block[] inputBlocks = page.getBlocks();
+        Block[] outputBlocks = new Block[projections.size()];
+
+        for (int projectionIndex = 0; projectionIndex < projections.size(); projectionIndex++) {
+            ProjectionFunction projection = projections.get(projectionIndex);
+
+            if (canDictionaryProcess(projection, inputPage)) {
+                Block outputDictionary = projectDictionary(projection, inputPage);
+                int[] outputIds = filterIds(projection, inputPage, selectedPositions);
+                outputBlocks[projectionIndex] = new DictionaryBlock(selectedPositions.length, outputDictionary, wrappedIntArray(outputIds));
+            }
+            else {
+                BlockBuilder blockBuilder = pageBuilder.getBlockBuilder(projectionIndex);
+                for (int position : selectedPositions) {
+                    projection.project(position, inputBlocks, blockBuilder);
+                }
+                outputBlocks[projectionIndex] = blockBuilder.build();
+            }
+        }
+
+        for (Block block : outputBlocks) {
+            verify(block.getPositionCount() == selectedPositions.length);
+        }
+        return new Page(outputBlocks);
+    }
+
+    private static int[] filterIds(ProjectionFunction projection, Page page, int[] selectedPositions)
+    {
+        Slice ids = ((DictionaryBlock) page.getBlock(getOnlyElement(projection.getInputChannels()))).getIds();
+
+        int[] outputIds = new int[selectedPositions.length];
+        for (int pos = 0; pos < selectedPositions.length; pos++) {
+            outputIds[pos] = ids.getInt(selectedPositions[pos] * SizeOf.SIZE_OF_INT);
+        }
+        return outputIds;
+    }
+
+    private static Block projectDictionary(ProjectionFunction projection, Page page)
+    {
+        int inputChannel = getOnlyElement(projection.getInputChannels());
+        Block dictionary = ((DictionaryBlock) page.getBlock(inputChannel)).getDictionary();
```

Comment:
On a per projection basis, we should store the last dictionary we processed and the result of the calculation.  Then we can short circuit the extra calculations.  This should be a big speed up for ORC files that use a single shared dictionary for the stripe.